### PR TITLE
plan, expression: support `is null` in hash partition pruning

### PR DESCRIPTION
**Pull request by imtbkcat**
### What problem does this PR solve?

Problem Summary:

SQL like `select * from t where a is null` will produce a full table scan plan in hash partition. For example:

```
create table t (a int, b int) partition by hash(a + b) partitions 10;
explain select * from t where a is null and b = 2;
```

This SQL will produce following result:

```
MySQL [test]> explain select * from t where a is null and b = 2;
+------------------------------+----------+-----------+-----------------------+-----------------------------------+
| id                           | estRows  | task      | access object         | operator info                     |
+------------------------------+----------+-----------+-----------------------+-----------------------------------+
| Union_16                     | 0.10     | root      |                       |                                   |
| ├─TableReader_19             | 0.01     | root      |                       | data:Selection_18                 |
| │ └─Selection_18             | 0.01     | cop[tikv] |                       | eq(test.t.b, 2), isnull(test.t.a) |
| │   └─TableFullScan_17       | 10000.00 | cop[tikv] | table:t, partition:p0 | keep order:false, stats:pseudo    |
| ├─TableReader_22             | 0.01     | root      |                       | data:Selection_21                 |
| │ └─Selection_21             | 0.01     | cop[tikv] |                       | eq(test.t.b, 2), isnull(test.t.a) |
| │   └─TableFullScan_20       | 10000.00 | cop[tikv] | table:t, partition:p1 | keep order:false, stats:pseudo    |
| ├─TableReader_25             | 0.01     | root      |                       | data:Selection_24                 |
| │ └─Selection_24             | 0.01     | cop[tikv] |                       | eq(test.t.b, 2), isnull(test.t.a) |
| │   └─TableFullScan_23       | 10000.00 | cop[tikv] | table:t, partition:p2 | keep order:false, stats:pseudo    |
| ├─TableReader_28             | 0.01     | root      |                       | data:Selection_27                 |
| │ └─Selection_27             | 0.01     | cop[tikv] |                       | eq(test.t.b, 2), isnull(test.t.a) |
| │   └─TableFullScan_26       | 10000.00 | cop[tikv] | table:t, partition:p3 | keep order:false, stats:pseudo    |
| ├─TableReader_31             | 0.01     | root      |                       | data:Selection_30                 |
| │ └─Selection_30             | 0.01     | cop[tikv] |                       | eq(test.t.b, 2), isnull(test.t.a) |
| │   └─TableFullScan_29       | 10000.00 | cop[tikv] | table:t, partition:p4 | keep order:false, stats:pseudo    |
| ├─TableReader_34             | 0.01     | root      |                       | data:Selection_33                 |
| │ └─Selection_33             | 0.01     | cop[tikv] |                       | eq(test.t.b, 2), isnull(test.t.a) |
| │   └─TableFullScan_32       | 10000.00 | cop[tikv] | table:t, partition:p5 | keep order:false, stats:pseudo    |
| ├─TableReader_37             | 0.01     | root      |                       | data:Selection_36                 |
| │ └─Selection_36             | 0.01     | cop[tikv] |                       | eq(test.t.b, 2), isnull(test.t.a) |
| │   └─TableFullScan_35       | 10000.00 | cop[tikv] | table:t, partition:p6 | keep order:false, stats:pseudo    |
| ├─TableReader_40             | 0.01     | root      |                       | data:Selection_39                 |
| │ └─Selection_39             | 0.01     | cop[tikv] |                       | eq(test.t.b, 2), isnull(test.t.a) |
| │   └─TableFullScan_38       | 10000.00 | cop[tikv] | table:t, partition:p7 | keep order:false, stats:pseudo    |
| ├─TableReader_43             | 0.01     | root      |                       | data:Selection_42                 |
| │ └─Selection_42             | 0.01     | cop[tikv] |                       | eq(test.t.b, 2), isnull(test.t.a) |
| │   └─TableFullScan_41       | 10000.00 | cop[tikv] | table:t, partition:p8 | keep order:false, stats:pseudo    |
| └─TableReader_46             | 0.01     | root      |                       | data:Selection_45                 |
|   └─Selection_45             | 0.01     | cop[tikv] |                       | eq(test.t.b, 2), isnull(test.t.a) |
|     └─TableFullScan_44       | 10000.00 | cop[tikv] | table:t, partition:p9 | keep order:false, stats:pseudo    |
+------------------------------+----------+-----------+-----------------------+-----------------------------------+
```

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

During partition pruning, if a column is equal to `NULL`, all calculation relating to it will produce null result and set a `NULL` flag.

How it Works:

If pruning result with a `NULL` flag, it will be point to `p0`.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

- No

### Release note 

- planner: support `is null` filter condition in hash partition pruning

### CODE REVIEW

File: expression/partition_pruner.go [hunk 1]
```diff
@@ -165,7 +182,7 @@ func newHashPartitionPruner() *hashPartitionPruner {
 
 // solve eval the hash partition expression, the first return value represent the result of partition expression. The second
 // return value is whether eval success. The third return value represent whether the eval result of partition value is null.
```
Comment (XuHuaiyu) [inline expression/partition_pruner.go:184]: 1. what does `conflict` mean?
2. The comment for the third value needs to be updated?

Comment (imtbkcat) [reply expression/partition_pruner.go:184]: conflict means condition like `a = 5 and a = 3` which will lead to filter condition to be false.

Comment (XuHuaiyu) [reply expression/partition_pruner.go:184]: maybe `isAlwaysFalse` is a better name

**Comment by tiancaiamao** [comment]
LGTM
Please fix CI

**Comment by imtbkcat** [comment]
@tiancaiamao @XuHuaiyu PTAL

File: expression/partition_pruner.go [hunk 2]
```diff
@@ -123,9 +134,11 @@ func (p *hashPartitionPruner) tryEvalPartitionExpr(piExpr Expression) (val int64
...
 			if val != nil {
 				pi.GetArgs()[0] = val
-				ret, _, err := pi.EvalInt(p.ctx, chunk.Row{})
+				ret, isNull, err := pi.EvalInt(p.ctx, chunk.Row{})
 				if err != nil {
 					return 0, false, false
+				} else if isNull {
+					return 0, true, true
 				}
 				return ret, true, false
```
Comment (XuHuaiyu) [inline expression/partition_pruner.go:143]: `return ret, true, isNull` is enough?

File: expression/testdata/partition_pruner_in.json [hunk 3]
```diff
@@ -2,23 +2,22 @@
...
       "explain select * from t2 join t1 on t1.id = t2.id and t2.a = t1.id and t2.id = 12",
-      // Negtive cases.
       "explain select * from t1 left join t2 on true where t1.a = 1 and false",
       "explain select * from t1 left join t2 on true where t1.a = 1 and null",
       "explain select * from t1 left join t2 on true where t1.a = null",
-      // Case with date.
       "explain select * from t4 where d = '2019-10-07 10:40:00' and a = 1",
-      "explain select * from t5 where d = '2019-10-07'"
+      "explain select * from t5 where d = '2019-10-07'",
+      "explain select * from t6 where a is null",
```
Comment (XuHuaiyu) [inline expression/testdata/partition_pruner_in.json:18]: add a case for `explain select * from t6 where b is null`

File: expression/partition_pruner.go (hunk 2 above)
Comment (imtbkcat) [reply expression/partition_pruner.go:143]: yes, for expression like `(year(null) + 2)` or `year(null)`, MySQL and TiDB both produce `NULL`

Comment (XuHuaiyu) [reply expression/partition_pruner.go:143]: what I mean is that: 
we can remove line 140~141, and change line 143 to `return ret, true, isNull`

File: expression/testdata/partition_pruner_in.json (hunk 3 above)
Comment (imtbkcat) [reply expression/testdata/partition_pruner_in.json:18]: done, PTAL

**Pull request merged**